paasta_tools/autoscaling_lib.py

```diff
@@ -12,97 +12,304 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from kazoo.client import KazooClient
-from service_configuration_lib import DEFAULT_SOA_DIR
+from datetime import datetime
+from math import ceil
 
+import requests
+from kazoo.exceptions import NoNodeError
+
+from paasta_tools.marathon_tools import compose_autoscaling_zookeeper_root
+from paasta_tools.marathon_tools import format_job_id
+from paasta_tools.marathon_tools import get_marathon_client
+from paasta_tools.marathon_tools import load_marathon_config
+from paasta_tools.marathon_tools import load_marathon_service_config
+from paasta_tools.marathon_tools import MESOS_TASK_SPACER
+from paasta_tools.marathon_tools import set_instances_for_marathon_service
+from paasta_tools.mesos_tools import get_running_tasks_from_active_frameworks
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import get_services_for_cluster
 from paasta_tools.utils import load_system_paasta_config
+from paasta_tools.utils import ZookeeperPool
+
+_autoscaling_ingesters = {}
+_autoscaling_deciders = {}
 
-_autoscaling_methods = {}
 
+def register_autoscaling_component(name, method_type):
+    if method_type == 'ingester':
+        component_dict = _autoscaling_ingesters
+    elif method_type == 'decider':
+        component_dict = _autoscaling_deciders
+    else:
+        raise KeyError('Autoscaling component type %s does not exist.' % method_type)
 
-def register_autoscaling_method(name):
     def outer(autoscaling_method):
-        _autoscaling_methods[name] = autoscaling_method
+        component_dict[name] = autoscaling_method
         return autoscaling_method
     return outer
 
 
-def get_autoscaling_method(function_name):
-    return _autoscaling_methods[function_name]
+def get_autoscaling_ingester(name):
+    return _autoscaling_ingesters[name]
 
 
-def compose_autoscaling_zookeeper_root(service, instance):
-    return '/autoscaling/%s/%s' % (service, instance)
+def get_autoscaling_decider(name):
+    return _autoscaling_deciders[name]
 
 
-def set_instances_for_marathon_service(service, instance, instance_count, soa_dir=DEFAULT_SOA_DIR):
-    zookeeper_path = '%s/instances' % compose_autoscaling_zookeeper_root(service, instance)
-    with ZookeeperPool() as zookeeper_client:
-        zookeeper_client.ensure_path(zookeeper_path)
-        zookeeper_client.set(zookeeper_path, str(instance_count))
+class IngesterNoDataError(ValueError):
+    pass
 
 
-def get_instances_from_zookeeper(service, instance):
-    with ZookeeperPool() as zookeeper_client:
-        (instances, _) = zookeeper_client.get('%s/instances' % compose_autoscaling_zookeeper_root(service, instance))
-        return int(instances)
+@register_autoscaling_component('threshold', 'decider')
+def threshold_decider(marathon_service_config, ingester_method, marathon_tasks, mesos_tasks,
```

Comment:
can you add a docstring to this function, and change the function to return True/False rather than 0/-1 ?